### PR TITLE
Create ImageAcquirer interface and package

### PR DESCRIPTION
**Pull request by enoodle**
This will add the `ImageAcquirer` interface to factor out the acquiring of the image from the main package. This encapsulation will enable easier testing and simpler code, and also will make it easier to add more ImageAcquirer in the future if we will need to.

The code for acquiring an Image from Docker's daemon was moved to `docker-image-acquirer.go` and to get an image from a running Docker container is in `docker-container-acquirer.go`.

I also creates the simpler `ScannerFactory` interface that `ImageInspector` implementation `defaultImageInspector` will implement. It is added for easier testing of ImageInspector as we already have the Scanner interface.

### CODE REVIEW

**Comment by enoodle** [comment]
cc @nimrodshn I will value your review as well

File: pkg/inspector/image-inspector_test.go [hunk 1]
```diff
@@ -1,215 +1,151 @@
...
+		opts.ScannerFactory = &happyEmptyScannerFactory{}
+	})
+	Describe("Inspect()", func() {
+		It("Simple Sanity with empty implementations", func() {
+			ii = NewDefaultImageInspector(*opts)
+			err = ii.Inspect()
+			Expect(err).NotTo(HaveOccurred())
+		})
+
+		It("quites when failing acquiring image", func() {
```
Comment (zgalor) [inline pkg/inspector/image-inspector_test.go]: quits

**Review by zgalor — APPROVED** [review]
some typo nits

File: pkg/imageacquirer/docker-image-acquirer.go [hunk 2]
```diff
@@ -0,0 +1,263 @@
...
+		return authCfgErr
+	}
+
+	// Try all the possible auth's from the config file
+	var err error
+	for name, auth := range imagePullAuths.Configs {
+		parsedErrors := make(chan error, 100)
+		defer func() { close(parsedErrors) }()
+
+		go func() {
```
Comment (nimrodshn) [inline pkg/imageacquirer/docker-image-acquirer.go]: @enoodle why do we put this block under a goroutine and then create another goroutine to decode the response?
Seem's to me like `parsedErrors` blocks anyways until `client.PullImage` and then `parsedError` blocks for `parsedErrors` so essentially the bottleneck is on `client.PullImage` So the following lines (the ones outside the goroutine) won't be running concurrently anyhow but after the goroutine is done.

Comment (enoodle) [reply pkg/imageacquirer/docker-image-acquirer.go]: @nimrodshn This code was moved from `pkg/inspector/image-inspector.go` and is not new here,
If I remember correctly this is done due to the error handling/writing to pipes and closing the other goroutines and channels. It is not and issue of bottle neck.
original PR: https://github.com/openshift/image-inspector/pull/42

Comment (nimrodshn) [reply pkg/imageacquirer/docker-image-acquirer.go]: Yeah I just looked it over and it looked wiered to me - Reading the comments in the PR it makes sense now 😃  👍

**Review by nimrodshn — APPROVED** [review]

**Comment by enoodle** [comment]
@simon3z @pweil- Can you take a look?

**Comment by simon3z** [comment]
> @simon3z @pweil- Can you take a look?

@enoodle sadly I don't have the time to review effectively a +988/−680 lines PR. Maybe you can try to get a thorough review from @pweil- or @mfojtik.

If it's possible to split this in multiple smaller PRs then generally that allows to get more attention on the isolated critical parts.

Overall the direction is good (adding the `ImageAcquirer` interface).

I gave a quick glimpse and I got some questions:

1. The method `Acquire(source string)` suggests that it's acquiring (download) the image but it's instead extracting it as well, so consider to either split this in two API calls `Acquire` and `Extract` (it seems nice for different use-cases). Or you if want to keep just a single API, then change the name to something more appropriate as `AcquireAndExtract`.
2. Why the `Acquire(source string)` code is also scanning the image? That leads to code duplication (each `ImageAcquirer` has to call the scanner with similar code), and also seems out of scope of the method given its name (see above).

**Comment by enoodle** [comment]
@simon3z 

> The method Acquire(source string) suggests that it's acquiring (download) the image but it's instead extracting it as well, so consider to either split this in two API calls Acquire and Extract (it seems nice for different use-cases). Or if you want to keep just a single API, then change the name to something more appropriate as AcquireAndExtract.

I wanted the interface to be simple, and I feel that extracting the image is an implementation detail (In the live container example it is not extracting anything)
I will change the function name and documentation to be clear that it does anything needed to get the image ready to be scanned and served.

> Why the Acquire(source string) code is also scanning the image? That leads to code duplication (each ImageAcquirer has to call the scanner with similar code), and also seems out of scope of the method given its name (see above).

This interface is meant to abstract the container runtime and the inspection is part of it. I don't see any value in separating it (again it could be that some runtimes will make it less relevant) but it can be done.
These inspection function just return metadata information about the image, they are not doing any deep inspection operations on them (the word "inspect" might be too overloaded here.)

There is no more code duplication with this interface than with the current implementation as most of the code was just moved to the new package.

**Comment by enoodle** [comment]
@simon3z This is my suggestion to refactor the `acquireImage` function [1]. It makes the addition of new runtimes much easier.

[1] https://github.com/openshift/image-inspector/pull/82#discussion_r163207801

cc @jhernand

**Pull request closed without merging**